Add animation for ErrorBar

ForestLinSen opened the pull request:
## Description

Added animation for ErrorBar component

## Related Issue

[ErrorBar does not animate](https://github.com/recharts/recharts/issues/4055)

## Motivation and Context

Previously the ErrorBar didn't have animation, now animation has been introduced to make it have a smoother visual effect

## How Has This Been Tested?

It was tested locally with Storybook and validated through unit tests

## Screenshots (if appropriate):

https://github.com/recharts/recharts/assets/41566276/32d7f85b-3e10-47de-87a8-ac5edd018072

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

ckifer commented on src/cartesian/ErrorBar.tsx:
```diff
@@ -130,5 +162,9 @@ ErrorBar.defaultProps = {
   width: 5,
   offset: 0,
   layout: 'horizontal',
+  isAnimationActive: true,
```
this is a breaking change (but this is the 3.x branch so that's ok)

PavelVanecek replied:
Why is it breaking? New feature, enabled by default, nothing breaks.

PavelVanecek replied:
We will have it in 3.x release anyway so ... it doesn't really matter. But I don't see it as breaking.

ckifer replied:
By breaking I mean it changed default behavior. If someone expects error bars to animate but on a patch release now they do, then I would consider that "breaking" in the sense that it should be defaulted to false until the next major release. In this case it doesn't matter though

ckifer commented on src/cartesian/ErrorBar.tsx:
```diff
@@ -43,12 +44,30 @@ interface ErrorBarProps extends InternalErrorBarProps {
    * Only accepts a value of "x" or "y" and makes the error bars lie in that direction.
    */
   direction?: 'x' | 'y';
+  isAnimationActive?: boolean;
+  animationBegin?: number;
+  animationDuration?: number;
+  animationEasing?: AnimationTiming;
```
can we add tests for each of the added properties?

1. animation can be disabled
2. animation starts after delay
3. animationDuration can be modified
4. animationEasing can be modified

ckifer reviewed (changes requested):
Looking for a few more tests here if possible.

I'd like to play with this a little before we release it to get the best possible default. From the video the appearance seems maybe not fast enough or maybe that the animation should follow the direction of the error bar? I'm not sure

Either way thanks for doing this

ForestLinSen commented:
> Looking for a few more tests here if possible.
> 
> I'd like to play with this a little before we release it to get the best possible default. From the video the appearance seems maybe not fast enough or maybe that the animation should follow the direction of the error bar? I'm not sure
> 
> Either way thanks for doing this

Thanks! I've updated the PR with more tests and also updated the default animation duration

PavelVanecek commented:
I wish the animation expanded from the dot up and down <3 that would be amazing.

PavelVanecek commented on test/cartesian/ErrorBar.spec.tsx:
```diff
@@ -103,4 +116,90 @@ describe('<ErrorBar />', () => {
+
+    assertErrorBars(container, 0);
+
+    setTimeout(() => {
```
If we use `setTimeout` in a test then failure from here will be reported in some completely different test in completely different file and mightily confuse everyone. Please use fake timers instead: https://vitest.dev/guide/mocking.html#timers

ForestLinSen replied:
thanks for the suggestion, tried to follow the document using `useFakeTimers` and `advanceTimersByTime` but failed to pass the test, could you give me any guidance on this?

```javascript
beforeAll(() => {
  vi.useFakeTimers();
});

afterAll(() => {
  vi.useRealTimers();
});

test('Renders Error Bars with animation', () => {
    const { container } = render(
      <LineChart data={barData} width={500} height={500}>
        <Line isAnimationActive dataKey="uv">
          <ErrorBar dataKey="uvError" />
        </Line>
      </LineChart>,
    );
      
    vi.advanceTimersByTime(200);
    assertErrorBars(container, 4);
    assertAnimationStyles(container, 'scale(1, 1)', 'transform 200ms ease-in-out 0s');
});
```

PavelVanecek commented on test/cartesian/ErrorBar.spec.tsx:
```diff
@@ -103,4 +116,90 @@ describe('<ErrorBar />', () => {
+    }, 1000);
+  });
+
+  // animationEasing
```
nit: this comment is not required, the information is in test title already (is it AI prompt? looks like AI prompt 🙃 )

ForestLinSen replied:
sry added four test names that ckifer mentioned there as todo reminders, forgot to delete the last one

PavelVanecek reviewed (changes requested):
Looks good to me except the setTimeout in test - please switch it to fake timers

ForestLinSen commented:
> I wish the animation expanded from the dot up and down <3 that would be amazing.

Did try that initially but couldn't find a way to implement that without using css @keyframe (I guess we shouldn't add extra css files in this project?)

ckifer commented:
@ForestLinSen looks like tests are failing

ForestLinSen commented:
> @ForestLinSen looks like tests are failing

Hi @ckifer , I've updated the unit tests, could you approve the workflow again? Thx

ckifer reviewed (approved):
Change LGTM

ckifer commented on test/cartesian/ErrorBar.spec.tsx:
```diff
@@ -103,4 +126,94 @@ describe('<ErrorBar />', () => {
   });
+
+  test('Renders Error Bars with animation', async () => {
+    mockAnimation();
```
afterthought and nit: but do the mock and the cleanup in a before/afterEach block here since you do it the same for each test

ForestLinSen replied:
good point! I've updated the code

ckifer reviewed (approved):
LGTM - going to merge it. If we want to adjust how the animation happens we probably can before 3.x is released

ckifer commented:
@ForestLinSen these tests failed after merge, could you take a look?
